Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs

```diff
@@ -188,6 +197,13 @@ public Task<TResult> InvokeWithParameterObjectAsync<TResult>(string targetName,
                 PythonConstants.ExtraPathsSetting, null, Site, PythonWorkspaceContextProvider.Workspace, out _)?.Split(';')
                 ?? _analysisOptions.ExtraPaths;
 
+            // Extra paths should include the project paths as well.
+            if (this.ProjectContextProvider != null) {
+                var directories = from n in this.ProjectContextProvider.ProjectNodes
+                                  select n.BaseURI.Directory;
+                extraPaths = extraPaths != null ? extraPaths.Concat(directories).ToArray() : directories.ToArray();
```

Comment:
@jakebailey would know more about extraPaths.

I'm thinking we shouldn't need to set this.

Reply:
This is required if these folders have to be import roots.

Better would be to simulate multi-root (if these folders really are distinct), but IDK what VS's status is on doing that.

Reply:
This was his recommendation actually.

Reply:
Otherwise it starts with just the root to the solution and doesn't find modules in the root or a project

Reply:
How does pylance see multiroot?

Reply:
Is there a new language server for each?

Reply:
I think this just needs an example project to show what you mean, maybe. Typically we don't want to blanket add each dir, but if these are folders containing the code that people edit and should be import roots, sure.

Reply:
Pylance supports multi-root natively, so only one LS process, but you have to have all of the calls for multi-root in LSP working.

Reply:
so just the base project root.. or something the user sets in the project settings as "root"

Reply:
In a solution there are multiple projects though. Each one of these needs to be parsed as a separate 'workspace' by Pylance, which I what I believe you're saying Jake?

Reply:
Here's a solution with multiple projects

![image](https://user-images.githubusercontent.com/19672699/112218865-7d3d3d80-8be1-11eb-9a7c-2cc7f21b8166.png)

Each 'project' here is like a workspace folder in VS code. This change is putting it into the 'extraPaths' setting. Is there another way to tell the LS about the workspace folders?

Reply:
The projects in the example above are:
- PythonApplication_Extra
- PythonApplication_Extra2
- PythonApplication2

Each of these has a different folder for its code. They may or may not be under the solution.

In that specific example, the solution is here:
C:\Users\aku91\source\repos\PythonApplication2

While the projects end up:
C:\Users\aku91\source\repos\PythonApplication2\PythonApplication_Extra
C:\Users\aku91\source\repos\PythonApplication2\PythonApplication_Extra2
C:\Users\aku91\source\repos\PythonApplication2\PythonApplication2

Although they don't have to be relative to the solution.

Reply:
I assume that you're talking about this notification instead?

https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_workspaceFolders

Reply:
yes, I believe this would be the better way to handle multiple projects, by using multiple workspaceFolders, instead of extrapaths.

Reply:
Yes, this seems much more like multi-root. See also https://microsoft.github.io/language-server-protocol/specification#initialize which has the folders field you could use.

Reply:
Okay thanks. I'm going to try getting that to work instead.